homeassistant/components/bring/todo.py

```diff
@@ -0,0 +1,183 @@
+
+from homeassistant import config_entries, core
+from homeassistant.components.todo import (
+    TodoItem,
+    TodoItemStatus,
+    TodoListEntity,
+    TodoListEntityFeature,
+)
+from homeassistant.core import callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN
+from .coordinator import BringData, BringDataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=90)
+
+
+async def async_setup_entry(
+    hass: core.HomeAssistant,
+    config_entry: config_entries.ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the sensor from a config entry created in the integrations UI."""
+    coordinator = hass.data[DOMAIN][config_entry.entry_id]
+
+    unique_id = config_entry.unique_id
+
+    if TYPE_CHECKING:
+        assert unique_id
+
+    for bringList in coordinator.data:
+        async_add_entities(
+            [
+                BringTodoListEntity(
+                    coordinator,
+                    bringList=bringList,
+                    unique_id=f"{unique_id} {bringList['listUuid']}",
+                )
+            ],
+        )
+
+
+class BringTodoListEntity(
+    CoordinatorEntity[BringDataUpdateCoordinator], TodoListEntity
+):
+    """A To-do List representation of the Bring! Shopping List."""
+
+    _attr_icon = "mdi:cart"
+    _attr_has_entity_name = True
+    _attr_supported_features = (
+        TodoListEntityFeature.CREATE_TODO_ITEM
+        | TodoListEntityFeature.UPDATE_TODO_ITEM
+        | TodoListEntityFeature.DELETE_TODO_ITEM
+        | TodoListEntityFeature.SET_DESCRIPTION_ON_ITEM
+    )
+
+    def __init__(
+        self,
+        coordinator: BringDataUpdateCoordinator,
+        bringList: BringData,
+        unique_id: str,
+    ) -> None:
+        """Initialize BringTodoListEntity."""
+        super().__init__(coordinator)
+        self.bringList = bringList
+        self._attr_name = self.bringList["name"]
+        self._attr_unique_id = unique_id
+
+    @property
+    def todo_items(self) -> list[TodoItem]:
+        """Return the todo items."""
+        return [
+            TodoItem(
+                uid=item["name"],
+                summary=item["name"],
+                description=item["specification"] or "",
+                status=TodoItemStatus.NEEDS_ACTION,
+            )
+            for item in self.bringList["items"]
+        ]
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator and update the bring list object."""
+        super()._handle_coordinator_update()
+        self.bringList = next(
+            (
+                lst
+                for lst in self.coordinator.data
+                if lst["listUuid"] == self.bringList["listUuid"]
+            ),
+        )
```

Comment:
switching the order of these 2 statements should fix the refresh problem

Reply:
Shouldn't the coordinator first refresh the data and then we get the the new list from him. If we switch it, in my understanding it should not work. Am I misunderstanding something? @joostlek

Reply:
I think the bring_list should be a separate `@property`. I haven't checked the full code, will do so in the train in about 30m

Reply:
Good point, that might do the job.

Reply:
Actually yes, the order was wrong and switching the order of the super() call and the bring list retrieval eliminated the "buggy" behaviour. But even better, converting the bringList into a property made even cleaner imo.